storage/replicate_queue.go

```diff
@@ -162,17 +165,31 @@ func (rq *replicateQueue) process(
 		log.Event(ctx, "removing a replica")
 		// We require the lease in order to process replicas, so
 		// repl.store.StoreID() corresponds to the lease-holder's store ID.
-		removeReplica, err := rq.allocator.RemoveTarget(desc.Replicas, repl.store.StoreID())
+		removeReplica, err := rq.allocator.RemoveTarget(desc.Replicas, 0)
 		if err != nil {
 			return err
 		}
-		log.VEventf(1, ctx, "removing replica %+v due to over-replication", removeReplica)
-		if err = repl.ChangeReplicas(ctx, roachpb.REMOVE_REPLICA, removeReplica, desc); err != nil {
-			return err
-		}
-		// Do not requeue if we removed ourselves.
 		if removeReplica.StoreID == repl.store.StoreID() {
-			return nil
+			var targetID roachpb.StoreID
+			for _, replica := range desc.Replicas {
```

Comment:
Choosing the first non-self replica as the target seems problematic. That store might be overloaded too, and could lead to ping-ponging between the first two stores in the list. We should at least randomize the choice, and ideally choose the most under-loaded one.